lib/src/front_end/comment_writer.dart

```diff
@@ -42,61 +41,43 @@ import 'piece_writer.dart';
 /// construct. These get directly embedded in the [TextPiece] of the code being
 /// written. When that [TextPiece] is output later, it will include the comments
 /// as well.
-mixin CommentWriter {
-  PieceWriter get pieces;
-
-  LineInfo get lineInfo;
+class CommentWriter {
+  final LineInfo _lineInfo;
 
   /// The tokens whose preceding comments have already been taken by calls to
   /// [takeCommentsBefore()].
   final Set<Token> _takenTokens = {};
 
+  CommentWriter(this._lineInfo);
+
   /// Returns the comments that appear before [token].
   ///
-  /// The caller is required to write them because a later call to [token()]
-  /// for this token will not write the preceding comments.
+  /// The caller is required to write them because a later call to write [token]
+  /// for this token will not write the preceding comments. Used by
+  /// [SequenceBuilder] and [DelimitedListBuilder] which handle comment
+  /// formatting themselves.
   CommentSequence takeCommentsBefore(Token token) {
     if (_takenTokens.contains(token)) return CommentSequence.empty;
     _takenTokens.add(token);
-    return _collectComments(token);
+    return _commentsBefore(token);
   }
 
-  /// Writes comments that appear before [token].
-  void writeCommentsBefore(Token token) {
+  /// Gets the comments that appear before [token].
```

Comment:
[nit] 
```suggestion
  /// Returns the comments that appear before [token].
```

Although the signature is clear, could also drop the doc.

Reply:
I'll apply this change locally.